src/sentry/api/endpoints/client_state.py

```diff
@@ -0,0 +1,52 @@
+# This endpoint helps managing persisted client state with a TTL for a member, organization or user
+
+from django.conf import settings
+from rest_framework.exceptions import NotFound
+from rest_framework.request import Request
+from rest_framework.response import Response
+
+from sentry.api.bases.organization import OrganizationEndpoint
+from sentry.utils import json, redis
+
+STATE_CATEGORIES = {
+    "onboarding": {
+        "ttl": 30 * 24 * 60 * 60,  # the time in seconds that the state will be persisted
+        "scope": "member",  # Can be "org" or "member"
+        "max_payload_size": 1024,
+    }
+}
+
+
+class ClientStateEndpoint(OrganizationEndpoint):
+    private = True
+
+    def __init__(self, **options) -> None:
+        cluster_key = getattr(settings, "SENTRY_CLIENT_STATE_REDIS_CLUSTER", "default")
+        self.client = redis.redis_clusters.get(cluster_key)
+        super().__init__(**options)
+
+    def get_key(self, organization, category, user):
+        if category not in STATE_CATEGORIES:
+            raise NotFound(detail="Category not found")
+        scope = STATE_CATEGORIES[category]["scope"]
+        if scope == "member":
+            return f"client-state:{category}:{organization}:{user}"
+        elif scope == "org":
+            return f"client-state:{category}:{organization}"
+
+    def get(self, request: Request, organization, category) -> Response:
+        key = self.get_key(organization.slug, category, request.user)
+        res = self.client.get(key)
+        if res:
+            self.client.expire(key, STATE_CATEGORIES[category]["ttl"])
+            return Response(json.loads(res))
```

Comment:
Why are you parsing JSON into a dict and then putting the dict into a Response that is going to reserialize the JSON?

Reply:
otherwise django-rest-framework will treat this as a string literal. I guess you can also pass in the serialized string and set the `Content-Type` header manually. Updated.